src/qutip_qip/decompositions/general_decompositions.py

```diff
@@ -0,0 +1,161 @@
+#
+#    THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+#    "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+#    LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A
+#    PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+#    HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+#    SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+#    LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+#    DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+#    THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+#    (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+#    OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+###############################################################################
+import numpy as np
+import cmath
+
+from qutip.qobj import Qobj
+
+class ConversionError():
+    pass
+
+
+def check_input(input_gate):
+    """Verifies input is a valid quantum gate.
+
+    Parameters
+    ----------
+    input_gate : :class:`qutip.Qobj`
+        The matrix that's supposed to be decomposed should be a Qobj.
+
+    Returns
+    -------
+    bool
+        If the input is a valid quantum gate matrix, the function returns "True".
+        In the case of "False" being returned, this function will ensure no
+        decomposition scheme can proceed.
+    """
+    # check if input is a qobj
+    qobj_check = isinstance(input_gate,Qobj)
+    if qobj_check == False:
+        raise TypeError("The input matrix is not a Qobj.")
+    else:
+        # check if input is square and a unitary
+        input_shape = input_gate.shape
+        if input_shape[0]==1:
+            raise ValueError("A 1-D Qobj is not a valid quantum gate.")
+        if input_shape[0] != input_shape[1]:
+            raise ValueError("Input is not a square matrix.")
+        else:
+            unitary_check = Qobj.check_isunitary(input_gate)
+            return unitary_check
+
+
+
+def find_qubits_in_circuit(input_gate):
+    """Based on the shape of the unitary input gate, determines the number of
+    qubits the input gate will act on.
+
+    TO DO: CHANGE THE LOOP FOR QUDITS (CURRENTLY ONLY WORKS FOR QUBITS)
+
+    Parameters
+    ----------
+    input_gate : :class:`qutip.Qobj`
+        The matrix that's supposed to be decomposed should be a Qobj.
+
+    Returns
+    -------
+    number_of_qubits: int
+        Returns an integer value for the number of qubits
+    """
+    input_check_bool = check_input(input_gate)
+    if input_check_bool == True:
+        input_shape = input_gate.shape
+        if input_shape[0]%2==0.0: # won't be 2 for higher qudits
+        # TO DO : if d=4 (qu4it) then it will still appear as a d=2 i.e. qubit gate
+        # change to powers of 2
+            number_of_qubits = np.log(input_shape[0])/np.log(2)
```

Comment:
@hodgestar @BoxiLi Is there a better way to find the number of qubits based on the size of the input gate matrix ? Here, with what I have chosen to do, if the qudits are of dimensions that are power of 2 then my condition is valid for both qubit and said qudit.

Reply:
@BoxiLi @hodgestar I edited my previous comment. I think I have a better idea now. It's possible my understanding of `dims` is wrong.

Reply:
Is there a reason that we need to consider qudits/qutrits? I thought the whole decomposition module assumes that we are working with qubits? Or is there anything also useful for qudits?

This kind of information is much easier to extract from the `Qobj`, before taking the full numpy matrix. A 2-qubit ket state has `dims=[[2, 2], [1, 1]]`, while a qudit ket `dims=[[4], [1]]`.

Reply:
For unitaries acting on them: `dims=[[2, 2], [2, 2]]` and `dims=[[4], [4]]`

Reply:
> Is there a reason that we need to consider qudits/qutrits?

Not really. As the user can input any n x n unitary, I added this extra check to make sure the unitary acts only on qubits and not on other dimensions.

Reply:
Ok, then this is just a validity check. Since the input is `Qobj`, using `dims` is easier. We can check if it is a list of `2`.
```
input_gate.dims[0] == [2] * len(input_gate.dims[0])
```

Reply:
Yeah, that's my problem. `dims` cannot distinguish between a 4 x 4 input. It could be an object acting on 2 qubits or a qudit of d = 4. 

```
q = Qobj([[0., 0, 0, 0.],[0., 0, 0, 0.],[0., 0, 0, 0.],[0., 0, 0, 0.]])
q.dims = [[4], [4]]
```
I think I am going to change the input to the function. In addition to the input Qobj, they will also need to input number of qubits. This could help avoid the ambiguity.

Reply:
Well, technically speaking, the example you provided is interpreted in QuTiP as a unitary acting on a *qudit* system, not two qubits.
E.g.
```
q = Qobj([[0., 0, 0, 0.],[0., 0, 0, 0.],[0., 0, 0, 0.],[0., 0, 0, 0.]], dims = [[4], [4]])
q * basis([2,2],[0,0])
```
will raise an error because the dimension does not match. This is how QuTiP distinguish them. If the user is meant to provide an operator acting on qubits, the dimension should match.

Reply:
That said, in decomposition we can also be lenient and allow the user to just produce a 2^n times 2^n `Qobj`.

So I see two choices:
- We make it strict and all the input must have the dimension `[[2,2,2,...],[2,2,2,...]`
- We allow it to be ambiguous a bit and require that `input.shape[0]` is a power of 2, by some smart bit manipulation: https://stackoverflow.com/questions/57025836/how-to-check-if-a-given-number-is-a-power-of-two